linkerd/app/inbound/src/policy/store.rs

```diff
@@ -84,53 +75,57 @@ impl Store {
+        tokio::spawn(async move {
+            default_tx.closed().await;
+        });
+
+        // The initial set of documented policies never expire from the cache.
+        //
+        // Policies that are dynamically discovered at runtime will expire after
+        // `idle_timeout` to prevent holding policy watches indefinitely for
+        // ports that are generally unused.
+        let cache = {
+            let rxs = ports.into_iter().map(|port| {
+                let watch = watch.clone();
+                let default = default_rx.borrow().clone();
+                let rx =
+                    info_span!("watch", %port).in_scope(|| watch.spawn_with_init(port, default));
                 (port, rx)
-            })
-            .collect::<PortMap<_>>();
-
-        let default = match Self::mk_default(default) {
-            Some((tx, rx)) => {
-                tokio::spawn(async move {
-                    tx.closed().await;
-                });
-                Some(rx)
-            }
-            None => None,
+            });
+            Cache::with_permanent_from_iter(idle_timeout, rxs)
         };
 
         Self {
-            default,
-            ports: Arc::new(rxs),
+            cache,
+            default_rx,
+            discover: Some(watch),
         }
     }
 }
 
-impl CheckPolicy for Store {
-    /// Checks that the destination port is configured to allow traffic.
-    ///
-    /// If the port is not explicitly configured, then the default policy is used. If the default
-    /// policy is `deny`, then a `DeniedUnknownPort` error is returned; otherwise an `AllowPolicy`
-    /// is returned that can be used to check whether the connection is permitted via
-    /// [`AllowPolicy::check_authorized`].
-    fn check_policy(&self, dst: OrigDstAddr) -> Result<AllowPolicy, DeniedUnknownPort> {
-        let server = self
-            .ports
-            .get(&dst.port())
-            .cloned()
-            .map(Ok)
-            .unwrap_or_else(|| match &self.default {
-                Some(rx) => Ok(rx.clone()),
-                None => Err(DeniedUnknownPort(dst.port())),
-            })?;
-
-        Ok(AllowPolicy { dst, server })
+impl<S> GetPolicy for Store<S>
+where
+    S: tonic::client::GrpcService<tonic::body::BoxBody, Error = Error>,
+    S: Clone + Send + Sync + 'static,
+    S::Future: Send,
+    S::ResponseBody:
+        http::HttpBody<Data = tonic::codegen::Bytes, Error = Error> + Default + Send + 'static,
+{
+    fn get_policy(&self, dst: OrigDstAddr) -> AllowPolicy {
+        // Lookup the polcify for the target port in the cache. If it doesn't
+        // already exist, we spawn a watch on the API (if it is configured). If
+        // no discovery API is configured we use the default policy.
+        let server =
+            self.cache
+                .get_or_insert_with(dst.port(), |port| match self.discover.clone() {
+                    Some(disco) => info_span!("watch", %port).in_scope(|| {
```

Comment:
since [`u16` implements `tracing::Value`](https://docs.rs/tracing/latest/tracing/trait.Value.html#impl-Value-for-u16), we don't need `Display` here
```suggestion
                    Some(disco) => info_span!("watch", port).in_scope(|| {
```